### PR TITLE
[develop] Add scaffolding to test login nodes dna_json

### PR DESCRIPTION
**Pull request by NSsirena**
### Description of changes
* Add scaffolding to test login nodes dna_json
* Add utility functions to render the user_data generated by CF and to check that it contains a set of fields expected in the dna.json file.
* Provide a simple test for dna.json fields.
* It may be used also for ComputeNodes.

### Tests
* Unit tests successfully executed locally.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/tests/pcluster/templates/test_cluster_stack.py [hunk 1]
```diff
@@ -767,6 +770,55 @@ def test_head_node_dna_json(mocker, test_datadir, config_file_name, expected_hea
...
+            {
+                "ephemeral_dir": "/scratch",
+                "enable_intel_hpc_platform": "false",
+                "head_node_private_ip": "{'Ref': 'referencetoclusternameHeadNodeENI6497A502PrimaryPrivateIpAddress'}",
+                "dns_domain": "{'Ref': 'referencetoclusternameClusterDNSDomain8D0872E1Ref'}",
+                "hosted_zone": "{'Ref': 'referencetoclusternameRoute53HostedZone2388733DRef'}",
+                "log_group_name": "/aws/parallelcluster/clustername",
+                "node_type": "LoginNode",
+                '"proxy"': "NONE",
+                "scheduler": "slurm",
```
Comment (jdeamicis) [inline cli/tests/pcluster/templates/test_cluster_stack.py]: Nitpick: maybe let's sort alphabetically.

File: cli/tests/pcluster/templates/test_cluster_stack.py [hunk 2]
```diff
@@ -767,6 +770,55 @@ def test_head_node_dna_json(mocker, test_datadir, config_file_name, expected_hea
...
+            },
+        ),
+    ],
+)
+def test_login_node_dna_json(mocker, test_datadir, config_file_name, expected_login_node_dna_json_fields):
+    mock_aws_api(mocker)
+    mock_bucket_object_utils(mocker)
+
+    # Read yaml and render CF Template
+    input_yaml = load_yaml_dict(test_datadir / config_file_name)
```
Comment (jdeamicis) [inline cli/tests/pcluster/templates/test_cluster_stack.py:804]: You haven't added the yaml file to the PR.

File: cli/tests/pcluster/utils.py [hunk 3]
```diff
@@ -65,6 +65,38 @@
     return None
 
 
+def get_resource_from_assets(assets, resource_name):
```
[the remark this answers is not in the document]
Comment (jdeamicis) [reply cli/tests/pcluster/utils.py:68]: Please fix this.

File: cli/tests/pcluster/utils.py [hunk 4]
```diff
@@ -65,6 +65,38 @@ def get_asset_content_with_resource_name(assets, resource_name):
...
+# 0: contains the file (as str) where those values will be replaced
+def render_user_data(user_data):
+    """Render the UserData as CF will do."""
+    content = user_data[0]
+    values = user_data[1]
+
+    for value in values:
+        content = content.replace("${%s}" % value, str(values[value]))
+
+    return content
```
Comment (jdeamicis) [inline cli/tests/pcluster/utils.py:88]: Isn't there a utility that can do this for us? Do we have to build and maintain this function ourselves?

File: cli/tests/pcluster/utils.py [hunk 5]
```diff
@@ -65,6 +65,38 @@ def get_asset_content_with_resource_name(assets, resource_name):
...
+
+
+def validate_dna_json_fields(user_data, fields):
```
Comment (jdeamicis) [inline cli/tests/pcluster/utils.py:119]: Can you please add a comment (or typing) to describe what the `user_data` variable should contain?

File: cli/tests/pcluster/utils.py [hunk 6]
```diff
@@ -65,6 +65,38 @@ def get_asset_content_with_resource_name(assets, resource_name):
...
+    """Validate that all dna.json fields, expressed as key-value pairs, are present in the user_data."""
+    lines = user_data.splitlines()
+    for k, v in fields.items():
+        for line in lines:
+            if k in line:
+                break
+
+        assert v in line, f"Validating Key '{k}' - Value: '{v}' not found in dna.json line:  '{line}'"
+
+
```
Comment (jdeamicis) [inline cli/tests/pcluster/utils.py:129]: Maybe I'm missing some context here, but why can't we import the rendered dna.json and the expected one as dictionaries and compare them via [`assertpy.is_subset_of`](https://assertpy.github.io/docs.html#assertpy.collection.CollectionMixin.is_subset_of)?

**Review by jdeamicis — CHANGES_REQUESTED** [review]
We must at least include the missing yaml in the PR.

I would have expected the rendering and the comparison functions to be simpler, but probably I'm missing something here.

File: cli/tests/pcluster/templates/test_cluster_stack.py (hunk 1 above)
Comment (NSsirena) [reply cli/tests/pcluster/templates/test_cluster_stack.py]: :+1: ..will do .. I picked in the order they are rendered in the dna_json ..but having them sorted would make it easier to debug.

File: cli/tests/pcluster/templates/test_cluster_stack.py (hunk 2 above)
Comment (NSsirena) [reply cli/tests/pcluster/templates/test_cluster_stack.py:804]: OMG !
... fixing it.

File: cli/tests/pcluster/utils.py (hunk 4 above)
Comment (NSsirena) [reply cli/tests/pcluster/utils.py:88]: I made a quick search and didn't find any...and then it was faster to write this function than searching further..

Happy to drop it if we have an alternative that we do not have to maintain...

File: cli/tests/pcluster/utils.py (hunk 5 above)
Comment (NSsirena) [reply cli/tests/pcluster/utils.py:119]: Will do

File: cli/tests/pcluster/utils.py (hunk 6 above)
Comment (NSsirena) [reply cli/tests/pcluster/utils.py:129]: For the HeadNode we are rendering the dna.json as a single file and so we can read it again as a dictionary.
For the LoginNodes (and ComputeNodes) the dna.json is a hidden script in the user_data that will create a file in the remote host while running.

Extracting this section from the user_data, creating a temporary file and re-reading it as a dictionary is something we can do ..but as a future improvement.

File: cli/tests/pcluster/utils.py (hunk 4 above)
Comment (NSsirena) [reply cli/tests/pcluster/utils.py:88]: As discussed in call we'll leave this as future improvement

File: cli/tests/pcluster/utils.py (hunk 5 above)
Comment (NSsirena) [reply cli/tests/pcluster/utils.py:119]: Done

**Review by jdeamicis — APPROVED** [review]

**Pull request merged**